fixed performance of agents

andrekassis opened the pull request:
In this commit we address the performance issue of agents using eBPF. We have come to realize that this mode of operation results in a very high CPU usage and we were able to attribute this phenomenon to the huge flow table that is kept in the memory and accessed using system calls which is a very expensive operation that slows down the system and results in the agents becoming stuck in the loop of trying to retrieve entries from the flow table. Hence, we suggest a fix in which we remove each entry after it has been read from the flow table and this way we are able to reduce the size of the table significantly eliminating the need to iterate over entries representing flows that had not been updated since we last encountered them.

In a future commit, we intend to address another serious issue. eBPF mode currently generates flow without updating their TCP metrics and therefore for such flows (TCP) we are unable to determine whether they have ended due to the lack of such info.
Cheers

safchain commented:
run tests

safchain commented:
ok to test

safchain commented:
run tests

safchain commented:
retest this please

safchain commented on flow/probes/ebpf.go:
```diff
@@ -207,7 +208,44 @@ func (p *EBPFProbe) flowFromEBPF(ebpfFlow *EBPFFlow, kernFlow *C.struct_flow, up
+                                                                 //Therefore, we remove every flow after extracting its data and we keep
+                                                                 //a local copy that we update continuously to preserve correctness.
+	if ebpfFlow.Kflow.link_layer._hash_src == kernFlow.link_layer._hash_src{
+		kernFlow.metrics.ab_packets+=ebpfFlow.Kflow.metrics.ab_packets
```
what about just ebpfFlow.Kflow.metrics.ab_packets+=... 

removing this line 
https://github.com/skydive-project/skydive/pull/1558/files#diff-733c12abf4ac0cf47dc4a9dddd3b9c40R320

andrekassis replied:
Took care of this

safchain commented on flow/probes/ebpf.go:
```diff
@@ -207,7 +208,44 @@ func (p *EBPFProbe) flowFromEBPF(ebpfFlow *EBPFFlow, kernFlow *C.struct_flow, up
+
+	if ebpfFlow.Kflow.metrics.ab_packets >= 1 && ebpfFlow.Kflow.metrics.ba_packets >= 1 && ebpfFlow.Kflow._flags != 1{
+		kernFlow.rtt+=kernFlow.last-ebpfFlow.Kflow.start
+		ebpfFlow.Kflow.rtt=kernFlow.rtt
```
same here, just ebpfFlow.Kflow.rtt = kernFlow.last-ebpfFlow.Kflow.start

andrekassis replied:
Same

safchain commented on flow/probes/ebpf.go:
```diff
@@ -271,21 +310,31 @@ func (p *EBPFProbe) run() {
 				}
+                                if ok{
+                                    updateFromEBPF(ebpfFlow,&kernFlow)
+                                    ebpfFlow.Kflow=kernFlow
```
seems to be useless ?

andrekassis replied:
should be the other way around after I added the above changes for we want to pass kernFlow to flowFromEBPF. Changed this as well

safchain commented on flow/probes/ebpf.go:
```diff
@@ -271,21 +293,31 @@ func (p *EBPFProbe) run() {
 				}
+                                if ok{
+                                    updateFromEBPF(ebpfFlow,&kernFlow)
+				    kernFlow=ebpfFlow.Kflow
```
I think you can remove this line and you can change the flowFromEBPF definition to remove the kernFlow parameter.

andrekassis replied:
Done

andrekassis commented on flow/probes/ebpf.go:
```diff
@@ -271,21 +293,30 @@ func (p *EBPFProbe) run() {
 					ebpfFlow.lastK = lastK
 					ebpfFlow.last = start.Add(time.Duration(lastK - startKTimeNs))
 
 					fl := p.flowFromEBPF(ebpfFlow, &kernFlow, unow)
```
done formatting. Check please?

safchain commented:
Can you please rebase and squash it on top of master we fixed the govet issue you are facing right now. Thanks. Otherwise LGTM

andrekassis commented:
Hi Sylvain
I've rebased and squashed on top of the master branch and the PR now is ready for you to merge.
I've also looked at your changes and they seem pretty fine to me.
I will run a few tests just to make sure everything is still functioning correctly and I'll let you know what I find.
Cheers;)

andrekassis commented:
Hi Sylvain,
I ran a few tests and your patch seems to work just fine;)
Best

safchain commented:
Hi Andre,

Thanks for testing :) did you see any improvement ? memory/cpu ?

andrekassis commented:
Hi Sylvain,
In answer to your question, I did not see any improvement but I'm sure that
this is just because of our tests that only generate very very small flows
so I'm confident that more advanced tests will be able to confirm your
changes do make a difference.
Andre

On Thu, Jan 17, 2019 at 5:22 PM Sylvain Afchain <notifications@github.com>
wrote:

> Hi Andre,
>
> Thanks for testing :) did you see any improvement ? memory/cpu ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/1558#issuecomment-455210192>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ArFJW2coM7qdkVKoaYnFXxOTlTIaQjkqks5vEJUzgaJpZM4Zk5MG>
> .
>
